tensorflow/contrib/layers/python/ops/layer_norm_fused_op.py

```diff
@@ -0,0 +1,90 @@
+# Copyright 2016 The TensorFlow Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Wrappers for efficient GPU layer_norm_fused operations."""
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+from tensorflow.contrib.util import loader
+from tensorflow.python.framework import ops
+from tensorflow.python.platform import resource_loader
+
+import tensorflow as tf
+import logging
+
+_layer_norm_fused_op = loader.load_op_library(
+    resource_loader.get_path_to_datafile("_layer_norm_fused_op.so"))
+
+
+@ops.RegisterGradient("LayerNormCustom")
+def _LayerNormCustomGrad(op, grad):
+    return [_layer_norm_fused_op.layer_norm_backprop_custom(
+        op.inputs[0], grad, op.get_attr("epsilon"))]
+
+
+@ops.RegisterGradient("LayerNormBiasAddCustom")
+def _LayerNormBiasAddCustomGrad(op, grad):
+    in_back, beta_back = _layer_norm_fused_op.layer_norm_bias_add_backprop_custom(
+        op.inputs[0], grad, op.inputs[1],
+        op.get_attr("epsilon"))
+    return [in_back, beta_back]
+
+
+@ops.RegisterGradient("LayerNormFusedCustom")
+def _LayerNormFusedCustomGrad(op, grad):
+    in_back, gamma_back, beta_back = _layer_norm_fused_op.layer_norm_fused_backprop_custom(
+        op.inputs[0], grad, op.inputs[1],
+        op.get_attr("epsilon"))
+    return [in_back, gamma_back, beta_back]
+
+
+def layer_norm_fused_op(input_tensor, gamma=None, beta=None,
+                        epsilon=1e-12, name=None):
+    """Fast and efficient layer normalization along the last dimension
```

Comment:
Yes, I also wondered whether I should add data_format as param as well, but then I realized NCHW/NHWC only applies to 4D tensors, and this op actually works on tensor rank >=2, and normalizes the last dimension regardless of tensor rank, so I thought data_format is less applicable here.